### PR TITLE
[loader]: Re-export symbols for C embedding, rename to `libjulia-internal`

### PR DESCRIPTION
**Pull request by staticfloat**
We first rename `libjulia` to `libjulia-internal`, and `libjulialoader`
to `libjulia` so as to reduce confusion from the outside; `libjulia` is
the entrypoint for all usecases, and `libjulia-internal` just so happens
to be the chunk that contains all the actual code and linkage to other
libraries.

To support C linking against `libjulia` in a natural way, we must
re-export our symbols from `libjulia-internal`.  In order to do this we
create function trampolines from `libjulia` to `libjulia-internal` and
we define all symbols within `libjulia` to be filled in by the
initialization routines of `libjulia-internal`.

The function and data symbols are now maintained in lists stored in
`src/jl_exported_*.list`.  These lists are consumed by `cli/Makefile`
where bindings are generated for them to be re-exported.  The function
trampolines are all of the form:

    __attribute__((visibility("hidden"))) void * jl_foo_addr;
    JL_DLLEXPORT JL_NAKED void jl_foo(void) { __asm__("jmpq *%0" : : "a"(jl_foo_addr) :); };

where the `jl_foo_addr` is initialized within `loader_export_symbols()`:

    jl_foo_addr = dlsym(libjulia_internal, "jl_foo");

The data symbols are merely defined as `void *` pointers within the
loader, then imported into `libjulia-internal` by the C linker.  This
backwards-dependency is slightly confusing, but it allows the symbols to
live within `libjulia` while being initialized by `libjulia-internal`.
At first glance it appears to be a circular dependency, however since
`libjulia` loads `libjulia-internal` through `dlopen()` first, then
`libjulia-internal` finds `libjulia` already loaded, the circular
dependency is a non-issue.  C code that wishes to link against
`libjulia` should do so directly, and the values of all symbols, while
initially being NULL, will be set to their correct values after
initialization through `load_libjulia_internal()` is complete.

### CODE REVIEW

**Comment by staticfloat** [comment]
@yuyichao I requested your review because I believe this was a central issue with regards to your objections to the `libjulialoader` PR.  With this change we should support direct C linking/embedding again.

@JeffBezanson I re-named `libjulia` to `libjulia-internal`, and `libjulialoader` to `libjulia`.  Let me know if there's something you don't like.

@vtjnash I think we're going to need to come up with function trampoline analogues that can work on any assembler.  So far it looks like `linux32` doesn't like `jmpq`.  Perhaps we should use `jmpl` on `x86`?

**Comment by yuyichao** [comment]
> With this change we should support direct C linking/embedding again.

No, as I said, the caller is still required to provide more information than before.
(and as an aside the tls handling is still wrong)

File: src/jl_exported_data.list [hunk 1]
```diff
@@ -0,0 +1,113 @@
+jl_abstractarray_type
```
Comment (vtjnash) [inline src/jl_exported_data.list:1]: Any particular reason not to make this a normal `.c` file? They're all just trivially `void *` variable decls.

File: cli/Makefile [hunk 2]
```diff
@@ -54,13 +53,34 @@ $(BUILDDIR)/julia_res.o: $(JULIAHOME)/contrib/windows/julia.rc $(JULIAHOME)/VERS
...
+	@cat $(SRCDIR)/../src/jl_exported_data.list | awk '{ print "JL_DLLEXPORT void * " $$1 ";" }' >> $@
+	@echo "// Function symbols" >> $@
+	@cat $(SRCDIR)/../src/jl_exported_funcs.list | awk '{ print "$(HIDDEN_SYMBOL) void * " $$1 "_addr;" }' >> $@
+	@echo >> $@
+	@echo "void loader_export_symbols(void * libjulia_internal) {" >> $@
+	@echo "    // Import function symbols" >> $@
+	@cat $(SRCDIR)/../src/jl_exported_funcs.list | awk '{ print "    " $$1 "_addr = dlsym(libjulia_internal, \"" $$1 "\");" }' >> $@
+	@echo "}" >> $@
+	@echo "// Define function trampolines" >> $@
+	@cat $(SRCDIR)/../src/jl_exported_funcs.list | awk '{ print "JL_DLLEXPORT JL_NAKED void " $$1 "(void) { __asm__(\"jmpq *%0\" : : \"a\"(" $$1 "_addr) :); };" }' >> $@
```
Comment (vtjnash) [inline cli/Makefile]: Since you're going to need a different macro for every platform, I'd suggest splitting this to a decl and an architecture #define:
```suggestion
	@echo "#define JL_DLLREEXPORT(f) JL_DLLEXPORT JL_NAKED void #f(void) { __asm__(\"jmpq *%0\" : : \"a\"#f#_addr : ); };" }' >> $@
	@cat $(SRCDIR)/../src/jl_exported_funcs.list | awk '{ print "JL_DLLREEXPORT(" $$1 ")" }' >> $@
```

File: src/jl_exported_data.list (hunk 1 above)
Comment (staticfloat) [reply src/jl_exported_data.list:1]: To keep symmetry with `jl_exported_funcs.list`; but it's not that big an issue.  It seemed nice to have the `src/Makefile` produce `.list` files, then have those `.list` files be processed into a `.c` file, in a symmetric fashion.

File: cli/Makefile [hunk 3]
```diff
@@ -54,13 +53,34 @@ $(BUILDDIR)/julia_res.o: $(JULIAHOME)/contrib/windows/julia.rc $(JULIAHOME)/VERS
...
+	@echo "// This list of symbols auto-generated by cli/Makefile and sourced from src/jl_exported_*.list; do not edit" > $@
+	@echo "// Define Data symbols" >> $@
+	@cat $(SRCDIR)/../src/jl_exported_data.list | awk '{ print "JL_DLLEXPORT void * " $$1 ";" }' >> $@
+	@echo "// Function symbols" >> $@
+	@cat $(SRCDIR)/../src/jl_exported_funcs.list | awk '{ print "$(HIDDEN_SYMBOL) void * " $$1 "_addr;" }' >> $@
+	@echo >> $@
+	@echo "void loader_export_symbols(void * libjulia_internal) {" >> $@
+	@echo "    // Import function symbols" >> $@
+	@cat $(SRCDIR)/../src/jl_exported_funcs.list | awk '{ print "    " $$1 "_addr = lookup_symbol(libjulia_internal, \"" $$1 "\");" }' >> $@
+	@echo "}" >> $@
```
Comment (vtjnash) [inline cli/Makefile:76]: To avoid the need for awk as a preprocessor, it's perhaps a bit more typically to use the c preprocessor (example adapted from uv.h. You can find similar in LLVM, though often with #include instead.):
```c
#define JL_FUNCTION_MAP(XX) \
  XX(jl_) \
  XX(jl_abs_float) \
  XX(jl_abs_float_withtype) \
  XX(jl_add_float) \
  XX(jl_add_int) \
  XX(jl_add_optimization_passes) \
...

#define STR(a) #a
#define XSTR(x) STR(x)

#if defined _CPU_X86_64_
#define JL_DLLREEXPORT(f) \
    JL_DLLEXPORT JL_NAKED void f(void) { __asm__(\"jmpq *%0\" : : \"a\"f#_addr : ) }
#elif defined _CPU_X86_
#define JL_DLLREEXPORT(f) \
    JL_DLLEXPORT JL_NAKED void f(void) { __asm__(\"jmpl *%0\" : : \"a\"f#_addr : ) }
#else
#warning "trampoline support needs to be defined for this CPU to enable linking against libjulia"
#undef JL_FUNCTION_MAP
#define JL_FUNCTION_MAP(x)
#endif

JL_FUNCTION_MAP(JL_DLLREEXPORT)
#undef JL_DLLREEXPORT

#define XX(f) static void *#f#_addr;
JL_FUNCTION_MAP(XX)
#undef XX

#define XX(f) XSTR(f),
static const char *const func_names[] = {
JL_FUNCTION_MAP(XX)
NULL
};
#undef XX

#define XX(f) &#f#_addr,
static void *const *const func_addrs[] = {
JL_FUNCTION_MAP(XX)
NULL
};
#undef XX

void loader_export_symbols(void * libjulia_internal)
{
    auto names = func_names;
    auto addrs = func_addrs;
    while (*names) {
        *addrs++ = dlsym(libjulia_internal, *names++);
    }
}
```

File: src/support/dtypes.h [hunk 4]
```diff
@@ -63,14 +63,16 @@
...
+#define JL_NAKED __attribute__ ((naked))
 # ifdef LIBRARY_EXPORTS
 #  define JL_DLLEXPORT __declspec(dllexport)
 # else
 #  define JL_DLLEXPORT __declspec(dllimport)
 # endif
 #else
 #define STDCALL
+#define JL_NAKED     __attribute__ ((naked))
 #define JL_DLLEXPORT __attribute__ ((visibility("default")))
```
Comment (vtjnash) [inline src/support/dtypes.h:76]:
```suggestion
#define JL_DLLEXPORT __attribute__ ((visibility("protected")))
```

Need to change this to make linux act like the other operating systems

File: src/jl_exported_data.list (hunk 1 above)
Comment (vtjnash) [reply src/jl_exported_data.list:1]: But seems like it's not producing `.list` files? And might be preferable to make `.inc` files (cf. LLVM) instead, so you don't need extra Makefile rules, but can directly #include them.

File: cli/Makefile (hunk 2 above)
Comment (staticfloat) [reply cli/Makefile]: Ah, I ended up creating a `JL_JUMPTO(x)` and using that instead.  I think that should be equivalent?

File: cli/loader.h [hunk 5]
```diff
@@ -49,16 +49,42 @@
...
+// Build assembler-dependent jump definitions
+#if defined(_CPU_X86_64_)
+#define JL_JUMP_TO(x)   __asm__("jmpq *%0" : : "a"(x) :)
+#elif defined(_CPU_X86_)
+#define JL_JUMP_TO(x)   __asm__("jmpl *%0" : : "a"(x) :)
+#elif defined(_CPU_AARCH64_)
+#define JL_JUMP_TO(x)   __asm__("br *%0" : : "a"(x) :)
+#elif defined(_CPU_ARM_)
+#define JL_JUMP_TO(x)   __asm__("br *%0" : : "a"(x) :)
+#elif defined(_ARCH_PPC)
```
Comment (vtjnash) [inline cli/loader.h:86]:
```suggestion
#elif defined(_ARCH_PPC_)
```

File: cli/loader.h [hunk 6]
```diff
@@ -49,16 +49,42 @@
...
+#define JL_JUMP_TO(x)   __asm__("ljmp *%0" : : "a"(x) :)
+#elif defined(__wasm__)
+#define JL_JUMP_TO(x)   __asm__("longjmp *%0" : : "a"(x) :)
+#endif
```
Comment (vtjnash) [inline cli/loader.h:90]:
```suggestion
#else
handle this with a warning or error
#endif
```

File: cli/loader.h [hunk 7]
```diff
@@ -49,16 +49,42 @@
...
 #endif
+
+
+// Build assembler-dependent jump definitions
+#if defined(_CPU_X86_64_)
+#define JL_JUMP_TO(x)   __asm__("jmpq *%0" : : "a"(x) :)
+#elif defined(_CPU_X86_)
+#define JL_JUMP_TO(x)   __asm__("jmpl *%0" : : "a"(x) :)
+#elif defined(_CPU_AARCH64_)
+#define JL_JUMP_TO(x)   __asm__("br *%0" : : "a"(x) :)
```
Comment (vtjnash) [inline cli/loader.h:83]: I don't think these other platforms have a `ax` register (the "a" constraint)

File: src/staticdata.c [hunk 8]
```diff
@@ -32,61 +32,174 @@ extern "C" {
...
-         &jl_interrupt_exception, &jl_boundserror_type, &jl_memory_exception,
-         &jl_undefref_exception, &jl_readonlymemory_exception,
+jl_value_t **const*const get_tags(void) {
+    // Make sure to keep an extra slot at the end to sentinel length
+    static void * _tags[146] = {NULL};
+
+    // Lazyily-initialize this list
+    if (_tags[0] == NULL) {
+        // builtin types
+        _tags[  0] = &jl_any_type;
```
Comment (c42f) [inline src/staticdata.c:42]: Just a drive-by comment — something like the following could make this a little neater, I guess?

```C
    const int ntags = 146;
    static void * _tags[ntags] = {NULL};
    if (_tags[0] == NULL) {
        int i = 0;
#define GVAR(sym) = _tags[i++] = &sym;
        GVAR(jl_any_type); GVAR(jl_symbol_type); ...
        // all globals ...

#undef GVAR
        assert(i == ntags);
```

Comment (staticfloat) [reply src/staticdata.c:42]: I liked this so much, I blindly used it, and got a little confused by the error messages until I saw that you have been writing a little too much Julia code.  ;)

```
#define GVAR(sym) = _tags[i++] = &(sym)
```

Comment (c42f) [reply src/staticdata.c:42]: Haha, I had to stare at this for a full 30 s until I saw that extra `=` sign.  You're right, too much Julia lately and no enough C :laughing:

**Comment by staticfloat** [comment]
This looks like it's good to go now!  Things that have changed since the last time you may have looked:

* The `embedding` tests work now!  There should be no extra information needed, we now use `dlpath()` equivalents to get the path of `libjulia.so` and build our relative paths to dependencies off of that.

* We don't rely on `__attribute__((naked))` anymore, but instead have manually-written assembly routines that essentially reimplement the PLT on all of our architectures.  This allows us to manually re-export symbols from `libjulia.so` to point to `libjulia-internal.so`, so that the C linker is happy to link against `libjulia.so`.

* We store all data symbols within `libjulia.so` and initialize their values using code in `libjulia-internal.so` invoked from things like `jl_init()`.

Note that we will most likely need to write a new assembly file for aarch64-darwin as it's not using `gas` as its assembler and the syntax may be a little different, but we can cross that bridge when we come to it.  As it stands, we have graceful degradation on new architectures which will be able to compile Julia just fine, but `libjulia.so` won't have the symbols that it needs for embedding on those architectures.  (For aarch64-darwin in particular we'll probably get a syntax error, since we already have a `aarch64` file).

File: .gitignore [hunk 9]
```diff
@@ -31,3 +31,4 @@
 /perf*
 .DS_Store
 .idea/*
+.vscode/*
```
Comment (vtjnash) [inline .gitignore:34]: These should be in your personal config

File: cli/loader_lib.c [hunk 10]
```diff
@@ -186,6 +193,9 @@ int load_repl(const char * exe_dir, int argc, char * argv[])
 // Empty DLL main entrypoint to silence warning
 #ifdef _OS_WINDOWS_
 int __stdcall DllMainCRTStartup(void* instance, unsigned reason, void* reserved) {
+    // Because we override DllMainCRTStartup, we have to manuall call our constructor methods
```
Comment (vtjnash) [inline cli/loader_lib.c:196]:
```suggestion
    // Because we override DllMainCRTStartup, we have to manually call our constructor methods
```

Comment (vtjnash) [reply cli/loader_lib.c:196]: Is there a particular reason we don't call CRT_INIT here as expected (such as not linking in that particular support library)?
https://docs.microsoft.com/en-us/troubleshoot/cpp/use-c-run-time

File: cli/trampolines/trampolines_aarch64.S [hunk 11]
```diff
@@ -0,0 +1,11 @@
+#include "../../src/jl_exported_funcs.inc"
+
+#define XX(name) \
+.global name; \
+name##:; \
```
Comment (vtjnash) [inline cli/trampolines/trampolines_aarch64.S:5]: @Keno do you happen to already know any good references of the `.cfi_startproc` and such that we should have here to preserve `@profile` capability when inside these frames. We can probably derive them by looking at the output of clang/gcc for these different targets, if not.

File: src/gf.c [hunk 12]
```diff
@@ -69,12 +69,18 @@ void jl_call_tracer(tracer_cb callback, jl_value_t *tracee)
...
-    1, &jl_method_type
-};
-static const struct jl_typemap_info lambda_cache = {
-    0, &jl_method_instance_type
-};
+static struct jl_typemap_info method_defs = {0};
+static struct jl_typemap_info lambda_cache = {0};
+
+// The Windows linker is unable to grab the imported `jl_method_type` and `jl_method_instance_type`
+// values at link-time, so we use a constructor method to do the importing ourselves.
```
Comment (vtjnash) [inline src/gf.c:76]: We don't have any actual use for these values, they're present as comments, so they can be turned into text

File: .gitignore (hunk 9 above)
Comment (staticfloat) [reply .gitignore:34]: I think we don't want to store `.vscode/*` from anybody, not just from me.

File: src/gf.c (hunk 12 above)
Comment (staticfloat) [reply src/gf.c:76]: I'm not sure what this means.

File: cli/loader_lib.c [hunk 13]
```diff
@@ -1,40 +1,36 @@
...
- * dummy value just so this file compiles on its own, and also so that developers can
- * see what this value should look like.  Note that the last entry must always be
- * `libjulia`, and that all paths should be relative to this loader `.exe` path.
- */
-#if !defined(DEP_LIBS)
-#define DEP_LIBS "../lib/example.so:../lib/libjulia.so"
-#endif
-static char dep_libs[256] = DEP_LIBS;
+// Save DEP_LIBS to a variable that is explicitly sized for expansion
+static char dep_libs[512] = DEP_LIBS;
```
[the remark this answers is not in the document]
Comment (staticfloat) [reply cli/loader_lib.c:18]: I don't see a way we can avoid mutating later; we need to be able to change these paths because the relative paths change between `make` time and `make install` time.

[the remark this answers is not in the document]
Comment (staticfloat) [reply cli/loader_lib.c:18]: Ah, I see, you're talking about the mutation later in this same file.  I can try to replace the `:` with `\0`, we'll see how it interacts with the rest of the processing of `DEPS_LIBS`, I have a feeling `stringreplace` isn't going to like this.

[the remark this answers is not in the document]
Comment (staticfloat) [reply cli/loader_lib.c:18]: Yeah, `stringreplace` doesn't want to deal with strings that have NULLs in them.  I'm going to leave this as-is for now.

File: cli/loader_lib.c (hunk 10 above)
Comment (staticfloat) [reply cli/loader_lib.c]: We don't link to that library at all, so I don't think we need to do this.

File: src/gf.c (hunk 12 above)
Comment (vtjnash) [reply src/gf.c:76]: You can just delete all this

File: .gitignore (hunk 9 above)
Comment (vtjnash) [reply .gitignore:34]: It's old project policy essentially. Hasn't come up in quite a long time surprisingly (other than https://github.com/JuliaLang/julia/pull/34169 sneaking .idea in?), per https://github.com/JuliaLang/julia/commit/f62ac13b42f11ca128117e77fab1632f7e9ac50c#diff-bc37d034bad564583790a46f19d807abfe519c5671395fd494d8cce506c42947

Comment (staticfloat) [reply .gitignore:34]: I think we're going to need to adjust that policy in the near future; there are too many VSCode extensions that will store state that other projects are going to want to embed in their `.vscode` folders that devs aren't going to want to put this in their global `.gitignore` files.

Comment (staticfloat) [reply .gitignore:34]: I'll remove this from this PR though, it's totally unrelated.

File: src/gf.c [hunk 14]
```diff
@@ -69,12 +69,8 @@ void jl_call_tracer(tracer_cb callback, jl_value_t *tracee)
 
 /// ----- Definitions for various internal TypeMaps ----- ///
 
-static const struct jl_typemap_info method_defs = {
-    1, &jl_method_type
-};
-static const struct jl_typemap_info lambda_cache = {
-    0, &jl_method_instance_type
-};
+static struct jl_typemap_info method_defs = {0};
```
Comment (vtjnash) [inline src/gf.c:72]:
```suggestion
static struct jl_typemap_info method_defs = {1};
```

**Comment by staticfloat** [comment]
Looks like this is ready to merge.  Keno and I confirmed that you can get unwind info from within a trampoline by segfaulting within the trampoline and seeing the correct unwind trace (despite the fact that the trampoline function itself was referred to as an unknown function).

**Comment by staticfloat** [comment]
Green flush!  Let's go.

**Comment by GunnarFarneback** [comment]
> The embedding tests work now! There should be no extra information needed, we now use dlpath() equivalents to get the path of libjulia.so and build our relative paths to dependencies off of that.

They work with Julia in the build tree but not with the nightly binaries. More specifically, running
```
cd /tmp
wget https://julialangnightlies-s3.julialang.org/bin/linux/x64/julia-latest-linux64.tar.gz
tar xvfz julia-latest-linux64.tar.gz
cd ~/julia/test/embedding/
mkdir build
JULIA=/tmp/julia-6ddc7f1ecc/bin/julia BIN=build make
./build/embedding
```
I get
```
ERROR: could not load library "/tmp/julia-6ddc7f1ecc/lib/julia/../bin/../lib/julia/sys.so"
/tmp/julia-6ddc7f1ecc/lib/julia/../bin/../lib/julia/sys.so: cannot open shared object file: No such file or directory
```

**Comment by nalimilan** [comment]
@staticfloat I get a similar error as @GunnarFarneback above when trying to build the Fedora RPM package for Julia 1.6.0-beta. Any idea about how we can fix out of tree builds? (More precisely, I get an error when trying to load openlibm.)

**Comment by staticfloat** [comment]
Milan, can you open a new issue with all the relevant info please?  Thanks!

**Comment by nalimilan** [comment]
Sure: see https://github.com/JuliaLang/julia/issues/39523.
